### ISSUE TITLE
after_connect not populating Topology.state.connection_pools in time

### ISSUE THREAD

**Issue by Sailias** [elixir-mongo/mongodb#158]
I'm using poolboy to start workers that make connections to Mongo.
If I send too many requests to poolboy and it spawns up new connections to Mongo, there is a timing issue where `state.connection_pools` in `Topology.ex` isn't populated in time because it receives messages out of order.

For example:

Mongo.ex
```
@spec start_link(Keyword.t) :: {:ok, pid} | {:error, Mongo.Error.t | term}
def start_link(opts) do
  res = Topology.start_link(opts)
  IO.inspect(res, label: "Started Mongo")
  res
end

@spec find(GenServer.server, collection, BSON.document, Keyword.t) :: cursor
def find(topology_pid, coll, filter, opts \\ []) do
  IO.inspect(topology_pid, label: "Finding")
  ...
end
```

Monitor.ex
```
@doc false
def connected(_connection, me, topology_pid) do
  IO.inspect(topology_pid, label: "Connected")
  GenServer.cast(topology_pid, {:connected, me})
end
```

Topology.ex
```
def handle_call({:connection, address}, _from, state) do
  IO.inspect(state.connection_pools, label: "Connection Pools #{inspect(self)}")
  {:reply, Map.fetch(state.connection_pools, address), state}
end
```

I will get output that looks like this:

```
Started Mongo: {:ok, #PID<0.2690.0>}
Finding: #PID<0.2690.0>
Connection Pools #PID<0.2690.0>: %{}
Connected: #PID<0.2690.0>
```

I believe the `after_connect` in DbConnection is handled after a pid is returned and poolboy starts using the pid before the `connected` is called in Topology which would result in an inbox for Topology looking like this:

```
:connection // called from connection_for_address
:connected
```

I recommend Monitor.init/1 be the following:

```
@doc false
  def init([server_description, topology_pid, heartbeat_frequency_ms, connection_opts]) do
    opts = # monitors don't authenticate and use the "admin" database
      connection_opts
      |> Keyword.put(:database, "admin")
      |> Keyword.put(:skip_auth, true)
      |> Keyword.put(:backoff_min, heartbeat_frequency_ms)
      |> Keyword.put(:backoff_max, heartbeat_frequency_ms)
      |> Keyword.put(:backoff_type, :rand)
      |> Keyword.put(:connection_type, :monitor)
      |> Keyword.put(:topology_pid, topology_pid)

    {:ok, pid} = DBConnection.start_link(Mongo.Protocol, opts)
    
    # Notify Topology that we are connected
    connected(pid, self(), topology_pid)
    
    :ok = GenServer.cast(self, :check)
    {:ok, %{
      connection_pid: pid,
      topology_pid: topology_pid,
      server_description: server_description,
      heartbeat_frequency_ms: heartbeat_frequency_ms,
      opts: opts
    }}
  end
```

Here we drop `after_connect` and call `connected` right away.

If you agree that I've identified the problem and accept the proposed solution, I can submit a pull request.

**Comment by ankhers** [comment]
I'm not sure if that is actually the issue. This driver currently works as follows:

1. Attempt to make a single connection to either the `:hostname` `:port` pair, or to each of the servers in the `:seeds` key. This uses `Mongo.Monitor`. It is meant for the driver to be able to communicate with the cluster for operations such as asking what that server believes the topology of the cluster looks like.
2. Once a successful `Monitor` connection has been made, it will attempt to make `Monitor` connections to all of the servers that that server knows about. It will also make the "real" connection(s) that your application will use in order to query Mongo to that one server.
3. Repeat step 2 until we are connected to every server in the cluster.

What looks like is happening in your case is you are querying the database before the "real" connections have been made. Which just so happens to also be before the `Monitor` process connects. I do not think that moving where we call `connected/3` will fix this issue. I believe it will also cause more issues. The reason I did this was because of #144, #142 and #141.

I am not really sure if there is anything that can really be done about this. Are you getting an error or anything? Maybe we can provide better error handling for events like this.

**Comment by Sailias** [comment]
@ankhers I agree that I am querying the database before the real connections have been made.  This is in line with what poolboy expects, that establishing a connection is synchronous which is why it's sending it more query requests under load.  (it returns a pid, and poolboy uses it)

The actual error I get is because `{:reply, Map.fetch(state.connection_pools, address), state}` returns `:error` because the address is not yet in `connection_pools`

I can test for `:error` as a return value and retry the query x number of times, or I can add a `Process.sleep(100)` after the `Mongo.start_link` to wait for it to be connected and thus have poolboy wait, but I feel those are both aren't ideal or documented.

I would think `Mongo.start_link` should wait until all connections have been established and are ready to use.  I don't know the code base that well, so I'm not sure what other problems my solution would introduce, but it was a way to set up the state prior to returning the pid.

**Comment by ankhers** [comment]
I think it would be better to look at why we seem to be passing `address` when we do not actually have a connection to that server yet.

**Comment by Sailias** [comment]
I still believe it's because the `init` function in `Monitor` returns successfully which propagates up through `Topology` and `Mongo` which return a valid `{:ok, pid}` when the connections haven't been established yet thus exposing it to premature queries.

The reason is because `DbConnection#after_connect` is a `handle_cast` and returns right away.
I feel that `Monitor` should wait until `connected` is called before completing the `init` function halting the `Mongo.start_link` until all servers are connected to.

I might be able to put together a sample project that demonstrates the error to assist validating it.  I believe I've given enough information to point in the right direction.

I do feel this is a critical bug that breaks the libraries base functionality.

**Comment by ankhers** [comment]
But if we call `connected` in `Monitor.init`, we are lying. The whole point of that function is to tell the `Topology` process that a connection to the server has been made successfully. I am open to looking into some sort of fix for this, but calling `connected` in `init` is not a fix.

**Comment by Sailias** [comment]
I have been looking at this code for only 4 hours in total so I am not in a position to make suggestions on the individual modules and their impact on the project as a whole. 

I am happy to provide a project that confirms the error to assist in a fix.

**Comment by Sailias** [comment]
Here is a repo that demonstrates the error:

https://github.com/Sailias/elixir_mongodb_error_app

**Comment by ankhers** [comment]
I apologize for the long delay in getting back to you on this.

Why are you creating a poolboy connection to the mongo connection? This is creating multiple beam processes to talk to a single database connection.

If you were to instead just start your mongo connection with `pool: DBConnection.Poolboy`, you would have multiple worker processes that actually all talk to the database.

As for running your code, I do not seem to get that error. Though I will continue to look into the reason behind it.

**Comment by Sailias** [comment]
I am spawning connections to new databases dynamically at runtime. Consider a multi tenant application where databases are dynamically created and configured in the application for each tenant. 

When the connection is established and queried immediately, the connection is not ready immediately. 

The repo I provided uses flow and a sleep to replicate the effects. 

I will try to run that repo using your latest version to see if there have been any changes since I created this issue.
